pyserini/search/__main__.py

```diff
@@ -130,37 +137,68 @@
 print(f'Running {args.topics} topics, saving to {output_path}...')
 tag = output_path[:-4] if args.output is None else 'Anserini'
 
-with open(output_path, 'w') as target_file:
-    for index, topic in enumerate(tqdm(sorted(topics.keys()))):
-        search = topics[topic].get('title')
-        hits = searcher.search(search, args.hits)
 
-        if not args.max_passage:
-            docids = [hit.docid.strip() for hit in hits]
-            scores = [hit.score for hit in hits]
+def write_result(result: Tuple[str, List[JSimpleSearcher]]):
+    topic, hits = result
+    docids = [hit.docid.strip() for hit in hits]
+    scores = [hit.score for hit in hits]
 
-            if use_prcl and len(hits) > (args.r + args.n):
-                scores, docids = ranker.rerank(docids, scores)
+    if use_prcl and len(hits) > (args.r + args.n):
+        scores, docids = ranker.rerank(docids, scores)
 
-            if args.msmarco:
-                for i, docid in enumerate(docids):
-                    target_file.write(f'{topic}\t{docid}\t{i + 1}\n')
-            else:
-                for i, (docid, score) in enumerate(zip(docids, scores)):
-                    target_file.write(f'{topic} Q0 {docid} {i + 1} {score:.6f} {tag}\n')
+    if args.msmarco:
+        for i, docid in enumerate(docids):
+            target_file.write(f'{topic}\t{docid}\t{i + 1}\n')
+    else:
+        for i, (docid, score) in enumerate(zip(docids, scores)):
+            target_file.write(
+                f'{topic} Q0 {docid} {i + 1} {score:.6f} {tag}\n')
+
+
+def write_result_max_passage(result: Tuple[str, List[JSimpleSearcher]]):
```

Comment:
This is actually more of a general Python question: How does this method get access to `target_file`, which is part of `with open(output_path, 'w') as target_file:` from the caller? It doesn't appear to be explicitly passed?

Reply:
In Python variables that are only referenced inside a function are [implicitly declared global](https://docs.python.org/3/faq/programming.html#id9). If you were to try assigning to `target_file` inside of `write_result` for example it would throw an error.